plugins/search-backend/src/service/AuthorizedSearchEngine.ts

```diff
@@ -0,0 +1,204 @@
+  constructor(
+    private readonly searchEngine: SearchEngine,
+    private readonly types: Record<string, DocumentTypeInfo>,
+    private readonly permissions: PermissionAuthorizer,
+    config: Config,
+  ) {
+    this.queryLatencyBudgetMs =
+      config.getOptionalNumber('search.permissions.queryLatencyBudgetMs') ??
+      1000;
+  }
+
+  setTranslator(translator: QueryTranslator): void {
+    this.searchEngine.setTranslator(translator);
+  }
+
+  async index(type: string, documents: IndexableDocument[]): Promise<void> {
+    this.searchEngine.index(type, documents);
+  }
+
+  async query(
+    query: SearchQuery,
+    options: QueryRequestOptions,
+  ): Promise<SearchResultSet> {
+    const queryStartTime = Date.now();
+
+    const authorizer = new DataLoader((requests: readonly AuthorizeQuery[]) =>
+      this.permissions.authorize(requests.slice(), options),
+    );
+    const requestedTypes = query.types || Object.keys(this.types);
+
+    const typeDecisions = zipObject(
+      requestedTypes,
+      await Promise.all(
+        requestedTypes.map(type => {
+          const permission = this.types[type]?.visibilityPermission;
+
+          return permission
+            ? authorizer.load({ permission })
+            : { result: AuthorizeResult.ALLOW as const };
+        }),
+      ),
+    );
+
+    const authorizedTypes = requestedTypes.filter(
+      type => typeDecisions[type]?.result !== AuthorizeResult.DENY,
+    );
+
+    const resultByResultFilteringRequired = authorizedTypes.some(
+      type => typeDecisions[type]?.result === AuthorizeResult.CONDITIONAL,
+    );
+
+    // When there are no CONDITIONAL decisions for any of the requested
+    // result types, we can skip filtering result by result by simply
+    // skipping the types the user is not permitted to see, which will
+    // be much more efficient.
+    //
+    // Since it's not currently possible to configure the page size used
+    // by search engines, this detail means that a single user might see
+    // a different page size depending on whether their search required
+    // result-by-result filtering or not. We can fix this minor
+    // inconsistency by introducing a configurable page size.
+    //
+    // cf. https://github.com/backstage/backstage/issues/9162
+    if (!resultByResultFilteringRequired) {
+      return this.searchEngine.query(
+        { ...query, types: authorizedTypes },
+        options,
+      );
+    }
+
+    const { page } = decodePageCursor(query.pageCursor);
+    const targetResults = (page + 1) * this.pageSize;
+
+    let filteredResults: SearchResult[] = [];
+    let nextPageCursor: string | undefined;
+    let latencyBudgetExhausted = false;
+
+    do {
+      const nextPage = await this.searchEngine.query(
+        { ...query, types: authorizedTypes, pageCursor: nextPageCursor },
+        options,
+      );
+
+      filteredResults = filteredResults.concat(
+        await this.filterResults(nextPage.results, typeDecisions, authorizer),
```

Comment:
What happens if the permissions service responds with an error?

Reply:
Sadly today, that error will propagate all the way back to the original /query call. In general resilience to errors is something we want to handle better in the permission-backend - right now a single failure in a batch of authorize queries will cause the whole batch to fail. 

We'd definitely like to adjust things so that it's possible for an individual decision to be replaced with an error and returned alongside successful responses, and once we do that we can improve the error handling in places like this, too.